libsolidity/lsp/LanguageServer.cpp

```diff
@@ -0,0 +1,966 @@
+
+	params["diagnostics"] = Json::arrayValue;
+	for (shared_ptr<Error const> const& error: m_compilerStack->errors())
+	{
+		// Don't show this warning. "This is a pre-release compiler version."
+		if (error->errorId().error == 3805)
+			continue;
+
+		SourceReferenceExtractor::Message const message = SourceReferenceExtractor::extract(*error);
+
+		Json::Value jsonDiag;
+		jsonDiag["source"] = "solc";
+		jsonDiag["severity"] = int(toDiagnosticSeverity(error->type()));
+		jsonDiag["message"] = message.primary.message;
+
+		if (message.errorId.has_value())
+			jsonDiag["code"] = Json::UInt64{message.errorId.value().error};
+
+		jsonDiag["range"]["start"] = toJson(LineColumn{message.primary.position.line, message.primary.startColumn});
+		jsonDiag["range"]["end"] = toJson(LineColumn{message.primary.position.line, message.primary.endColumn});
+
+		for (SourceReference const& secondary: message.secondary)
+		{
+			Json::Value jsonRelated;
+			jsonRelated["message"] = secondary.message;
+			jsonRelated["location"]["uri"] = toFileURI(secondary.sourceName);
+			jsonRelated["location"]["range"]["start"] = toJson(LineColumn{secondary.position.line, secondary.startColumn});
+			jsonRelated["location"]["range"]["end"] = toJson(LineColumn{secondary.position.line, secondary.endColumn});
+			jsonDiag["relatedInformation"].append(jsonRelated);
+		}
+
+		params["diagnostics"].append(jsonDiag);
+	}
+
+	m_client->notify("textDocument/publishDiagnostics", params);
+}
+
+frontend::ASTNode const* LanguageServer::findASTNode(LineColumn _position, std::string const& _fileName)
+{
+	if (!m_compilerStack)
+		return nullptr;
+
+	frontend::SourceUnit const& sourceUnit = m_compilerStack->ast(_fileName);
+	auto const sourcePos = sourceUnit.location().source->translateLineColumnToPosition(_position.line, _position.column);
+	if (!sourcePos.has_value())
+		return nullptr;
+
+	ASTNode const* closestMatch = ASTNodeLocator::locate(sourcePos.value(), sourceUnit);
+
+	if (!closestMatch)
+		trace(
+			"findASTNode not found for "s +
+			to_string(sourcePos.value()) + ":" +
+			to_string(_position.line) + ":" +
+			to_string(_position.column)
+		);
+	else
+		trace(
+			"findASTNode found for "s +
+			to_string(sourcePos.value()) + ":" +
+			to_string(_position.line) + ":" +
+			to_string(_position.column) + ": " +
+			closestMatch->location().text() + " (" +
+			typeid(*closestMatch).name() +
+			")"s
+		);
+
+
+	return closestMatch;
+}
+
+optional<SourceLocation> LanguageServer::declarationPosition(frontend::Declaration const* _declaration)
+{
+	if (!_declaration)
+		return nullopt;
+
+	if (_declaration->nameLocation().isValid())
+		return _declaration->nameLocation();
+
+	if (_declaration->location().isValid())
+		return _declaration->location();
+
+	return nullopt;
+}
+
+void LanguageServer::findAllReferences(
+	frontend::Declaration const* _declaration,
+	string const& _sourceIdentifierName,
+	frontend::SourceUnit const& _sourceUnit,
+	std::vector<SourceLocation>& _output
```

Comment:
Why not return the vector?

Reply:
Actually, I was in-place constructing the output vector, as it was in some cases called multiple times with a different Declaration pointer, that should go into the same output. I think performance may not matter too much here though. thanks.